### PR TITLE
Allow the Liveness Module's threshold to be customized

### PR DESCRIPTION
**Pull request by maurelian**
**Description**

Enables the required threshold of the Liveness Module to be set as an immutable in the constructor. This approach minimize config complexity. 

We need to be able to modify the required threshold so that we can operate the same modules in production on a testnet, without having to keep a 75% threshold which is excessive overhead.

## Summary by CodeRabbit

- **New Features**
	- Enhanced Safe contract security with a dynamic threshold percentage calculation.
	- Improved user transparency by adding threshold percentage retrieval and fallback owner functionality.
- **Refactor**
	- Updated LivenessModule contract logic for flexible threshold calculation based on the number of owners.

### CODE REVIEW

File: packages/contracts-bedrock/src/Safe/LivenessModule.sol [hunk 1]
```diff
@@ -47,25 +50,27 @@ contract LivenessModule is ISemver {
...
         SAFE = _safe;
         LIVENESS_GUARD = _livenessGuard;
         LIVENESS_INTERVAL = _livenessInterval;
+        THRESHOLD_PERCENTAGE = _thresholdPercentage;
         FALLBACK_OWNER = _fallbackOwner;
         MIN_OWNERS = _minOwners;
         address[] memory owners = _safe.getOwners();
         require(_minOwners <= owners.length, "LivenessModule: minOwners must be less than the number of owners");
         require(
-            _safe.getThreshold() >= get75PercentThreshold(owners.length),
```
Comment (tynes) [inline packages/contracts-bedrock/src/Safe/LivenessModule.sol:66]: I notice this doesn't bump semver, should be `1.1.0`?

**Comment by tynes** [comment]
How is this deployed?

**Comment by maurelian** [comment]
> How is this deployed?

@tynes By hand is my thinking. It only needs to be deployed once per Security Council.

**Comment by tynes** [comment]
Failing test

```
Failing tests:
Encountered 1 failing test in test/Safe/LivenessModule.t.sol:LivenessModule_Getters_Test
[FAIL. Reason: assertion failed] test_getters_works() (gas: 30308)
```

File: packages/contracts-bedrock/test/Safe/LivenessModule.t.sol [hunk 2]
```diff
@@ -310,9 +320,9 @@ contract LivenessModule_RemoveOwners_Test is LivenessModule_TestInit {
     }
 }
 
-/// @dev A copy of LivenessModule.get75PercentThreshold as a free function to use below.
-function get75PercentThreshold(uint256 _numOwners) pure returns (uint256 threshold_) {
-    threshold_ = (_numOwners * 75 + 99) / 100;
+/// @dev A copy of LivenessModule.getRequiredThreshold as a free function to use below.
+function getRequiredThreshold(uint256 _numOwners, uint256 _percentage) pure returns (uint256 threshold_) {
+    threshold_ = (_numOwners * _percentage + 99) / 100;
```
Comment (mds1) [inline packages/contracts-bedrock/test/Safe/LivenessModule.t.sol]: Seems safer/more maintainable to just deploy a dummy version of the module and call the public function, instead of copy/pasting the method?

File: packages/contracts-bedrock/test/Safe/LivenessModule.t.sol [hunk 3]
```diff
@@ -123,30 +133,30 @@ contract LivenessModule_CanRemove_Test is LivenessModule_TestInit {
...
-        assertEq(livenessModule.get75PercentThreshold(7), 6);
-        assertEq(livenessModule.get75PercentThreshold(6), 5);
-        assertEq(livenessModule.get75PercentThreshold(5), 4);
-        assertEq(livenessModule.get75PercentThreshold(4), 3);
-        assertEq(livenessModule.get75PercentThreshold(3), 3);
-        assertEq(livenessModule.get75PercentThreshold(2), 2);
-        assertEq(livenessModule.get75PercentThreshold(1), 1);
+contract LivenessModule_GetRequiredThreshold_Test is LivenessModule_TestInit {
+    /// @dev check the return values of the getRequiredThreshold function against manually
+    ///      calculated values, with 75% threshold.
```
Comment (mds1) [inline packages/contracts-bedrock/test/Safe/LivenessModule.t.sol]: What do you think about another set of hardcoded tests using a value other than 75%, and then a few using the 0% and 100% threshold bounds to test edge cases?

File: packages/contracts-bedrock/src/Safe/LivenessModule.sol [hunk 4]
```diff
@@ -47,25 +50,27 @@ contract LivenessModule is ISemver {
...
+            "LivenessModule: Insufficient threshold for the number of owners"
         );
     }
 
-    /// @notice For a given number of owners, return the lowest threshold which is greater than 75.
+    /// @notice For a given number of owners, return the lowest threshold which is greater than the required percentage.
     ///         Note: this function returns 1 for numOwners == 1.
-    function get75PercentThreshold(uint256 _numOwners) public pure returns (uint256 threshold_) {
-        threshold_ = (_numOwners * 75 + 99) / 100;
+    function getRequiredThreshold(uint256 _numOwners, uint256 _percentage) public pure returns (uint256 threshold_) {
```
Comment (mds1) [inline packages/contracts-bedrock/src/Safe/LivenessModule.sol]: Invariant is that all calls to `getRequiredThreshold` must pass in `THRESHOLD_PERCENTAGE` as the second arg, and that's a bit hard to test directly. Maybe we should just keep it as one argument like before?

**Comment by tynes** [comment]
The deploy script should be updated so we have a standard and public way to deploy this contract. It doesn't need to be part of the deploy the whole system flow

File: packages/contracts-bedrock/src/Safe/LivenessModule.sol (hunk 4 above)
Comment (maurelian) [reply packages/contracts-bedrock/src/Safe/LivenessModule.sol]: agree.

File: packages/contracts-bedrock/src/Safe/LivenessModule.sol [hunk 5]
```diff
@@ -29,6 +29,9 @@ contract LivenessModule is ISemver {
     ///         This can be updated by replacing with a new module.
     uint256 internal immutable MIN_OWNERS;
 
+    /// @notice The percentage used to calculate the threshold for the Safe.
+    uint256 internal immutable THRESHOLD_PERCENTAGE;
```
Comment (smartcontracts) [inline packages/contracts-bedrock/src/Safe/LivenessModule.sol:39]: Why percentage vs a specific number of owners?

Comment (maurelian) [reply packages/contracts-bedrock/src/Safe/LivenessModule.sol:39]: Because the Stage 1 requirement is 75%. This allows us to increase or decrease the number of owners without changing the module.

Comment (smartcontracts) [reply packages/contracts-bedrock/src/Safe/LivenessModule.sol:39]: Gotcha

File: packages/contracts-bedrock/src/Safe/LivenessModule.sol (hunk 4 above)
Comment (smartcontracts) [reply packages/contracts-bedrock/src/Safe/LivenessModule.sol]: +1 to this

**Review by smartcontracts — COMMENTED** [review]
Seems pretty good to me, will re-review once the comments from @mds1 are addressed

File: packages/contracts-bedrock/test/Safe/LivenessModule.t.sol (hunk 2 above)
Comment (maurelian) [reply packages/contracts-bedrock/test/Safe/LivenessModule.t.sol]: OK, that can be found here: 

https://github.com/ethereum-optimism/optimism/blob/maur/flexible-liveness-threshold/packages/contracts-bedrock/test/Safe/LivenessModule.t.sol#L49

File: packages/contracts-bedrock/test/Safe/LivenessModule.t.sol (hunk 3 above)
Comment (maurelian) [reply packages/contracts-bedrock/test/Safe/LivenessModule.t.sol]: Instead of doing that, I wrote an alternative implementation of the function being tested and wrote a differential fuzz test. 

I think it's a lot cleaner, LMK what you think.

https://github.com/ethereum-optimism/optimism/blob/maur/flexible-liveness-threshold/packages/contracts-bedrock/test/Safe/LivenessModule.t.sol#L158-L175

File: packages/contracts-bedrock/src/Safe/LivenessModule.sol (hunk 4 above)
Comment (maurelian) [reply packages/contracts-bedrock/src/Safe/LivenessModule.sol]: `THRESHOLD_PERCENTAGE` is now an immutable var.

**Comment by maurelian** [comment]
> The deploy script should be updated so we have a standard and public way to deploy this contract. It doesn't need to be part of the deploy the whole system flow

@tynes good call out. I have a WIP branch stacked on this one to do so, but I don't want to expand this PR too much more.

File: packages/contracts-bedrock/test/Safe/LivenessModule.t.sol [hunk 6]
```diff
@@ -20,9 +20,10 @@ contract LivenessModule_TestInit is Test, SafeTestTools {
     event RemovedOwner(address indexed owner);
     event OwnershipTransferredToFallback();
 
-    uint256 initTime = 10;
-    uint256 livenessInterval = 30 days;
-    uint256 minOwners = 6;
+    uint256 constant INIT_TIME = 10;
+    uint256 constant LIVENESS_INTERVAL = 30 days;
+    uint256 constant MIN_OWNERS = 6;
+    uint256 constant THRESHOLD_PERCENTAGE = 75;
```
[the remark this answers is not in the document]
Comment (maurelian) [reply packages/contracts-bedrock/test/Safe/LivenessModule.t.sol:26]: It's variable in the large fuzz test at the end of this file. Making it constant here is consistent with a unit testing approach.

File: packages/contracts-bedrock/test/Safe/LivenessModule.t.sol [hunk 7]
```diff
@@ -40,14 +41,35 @@ contract LivenessModule_TestInit is Test, SafeTestTools {
...
+        // after it is created so the constructor arguments are not important.
+        LivenessModule dummyLivenessModule = new LivenessModule({
+            _safe: Safe(payable(address(0))),
+            _livenessGuard: LivenessGuard(address(0)),
+            _livenessInterval: 0,
+            _thresholdPercentage: _thresholdPercentage,
+            _minOwners: 0,
+            _fallbackOwner: address(0)
+        });
+        return dummyLivenessModule.getRequiredThreshold(_numOwners);
```
Comment (maurelian) [inline packages/contracts-bedrock/test/Safe/LivenessModule.t.sol:65]: I was asked to make this change instead of copy pasting the code.

File: packages/contracts-bedrock/test/Safe/LivenessModule.t.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (maurelian) [reply packages/contracts-bedrock/test/Safe/LivenessModule.t.sol:26]: it is allowed to vary in the two tests which are most relevant to it: 
1. `testFuzz_removeOwners`
2. `testFuzz_getRequiredThreshold_works`

**Comment by maurelian** [comment]
* **#10224** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/10224?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#10207** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/10207?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#10120** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/10120?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#10072** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/10072?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @maurelian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/contracts-bedrock/test/Safe/LivenessModule.t.sol [hunk 8]
```diff
@@ -126,30 +153,36 @@ contract LivenessModule_CanRemove_Test is LivenessModule_TestInit {
...
+        // If the total multiplied by the percentage is not divisible by 100, we need to add 1 to the result to
+        // compensate for the rounding down by integer division.
+        if ((_total * _percentage) % 100 > 0) {
+            toAdd = 1;
+        }
+        return (_total * _percentage) / 100 + toAdd;
+    }
+
+    /// @dev Differentially tests the getRequiredThreshold function against _getLeastIntegerValueAbovePercentage
+    function testFuzz_getRequiredThreshold_works(uint256 _numOwners, uint256 _percentage) external {
```
Comment (smartcontracts) [inline packages/contracts-bedrock/test/Safe/LivenessModule.t.sol:178]: I liked the assertions for specific known values, especially if we can use values that would be reasonable for the SC safe to have. I'm pretty confident this code is correct and the fuzz is good but a few tests for known-good values would add extra safety.

Comment (maurelian) [reply packages/contracts-bedrock/test/Safe/LivenessModule.t.sol:178]: OK, I've restored the hardcoded values, added another set of them at 33%, as well as a new fuzz tests for 1 and 100% thresholds.

a32b0c80a87427950864c3cd414f03ad3f55db19

**Review by smartcontracts — APPROVED** [review]
